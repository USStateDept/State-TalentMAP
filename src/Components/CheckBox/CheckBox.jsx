import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { EMPTY_FUNCTION } from '../../Constants/PropTypes';
import { formatIdSpacing } from '../../utilities';

class CheckBox extends Component {
  constructor(props) {
    super(props);
    this.state = {
      checked: { value: this.props.value },
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.value !== this.props.value) {
      this.setState({ checked: { value: nextProps.value } });
    }
  }

  onCheck() {
    const { checked } = this.state;
    checked.value = !checked.value;
    this.setState({ checked },
      this.props.onCheckBoxClick(checked.value, { ...this.props }),
    );
  }

  render() {
    const { id, label, title, name, labelSrOnly, small, className, disabled } = this.props;
    const { checked } = this.state;
    const formattedId = formatIdSpacing(id);
    return (
      <div className={`usa-grid-full ${className} tm-checkbox ${small ? 'tm-checkbox-small' : ''}`}>
        <input
          type="checkbox"
          id={formattedId}
          title={title}
          name={name}
          value={checked.value}
          onChange={() => this.onCheck()}
          checked={checked.value}
          disabled={disabled}
        />
        <label htmlFor={formattedId}><span className={`${labelSrOnly ? 'usa-sr-only' : ''}`}>{label}</span></label>
      </div>
    );
  }
}

CheckBox.propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.node.isRequired,
  title: PropTypes.string,
  name: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]).isRequired,
  onCheckBoxClick: PropTypes.func,
  labelSrOnly: PropTypes.bool,
  small: PropTypes.bool,
  className: PropTypes.string,
  disabled: PropTypes.bool,
};

CheckBox.defaultProps = {
  title: '',
  name: '',
  onCheckBoxClick: EMPTY_FUNCTION,
  labelSrOnly: false,
  small: false,
  className: '',
  disabled: false,
};

export default CheckBox;
