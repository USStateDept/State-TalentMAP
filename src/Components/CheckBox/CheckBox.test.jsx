import { shallow } from 'enzyme';
import React from 'react';
import toJSON from 'enzyme-to-json';
import sinon from 'sinon';
import CheckBox from './CheckBox';

describe('CheckBoxComponent', () => {
  it('can receive props', () => {
    const wrapper = shallow(
      <CheckBox
        id="id"
        label="label"
        title="title"
        name="name"
        value="value"
        onCheckBoxClick={() => {}}
      />,
    );
    expect(wrapper.instance().props.id).toBe('id');
  });

  it('can call the onCheck function upon selection', () => {
    const onCheckBoxClick = sinon.spy();
    const wrapper = shallow(
      <CheckBox
        id="id"
        label="label"
        title="title"
        name="name"
        value="value"
        onCheckBoxClick={onCheckBoxClick}
      />,
    );

    wrapper.instance().onCheck();
    expect(wrapper.instance().state.checked.value).toBe(false);
    expect(onCheckBoxClick.calledOnce).toBe(true);
  });

  it('can apply the base class', () => {
    const wrapper = shallow(
      <CheckBox
        id="id"
        label="label"
        title="title"
        name="name"
        value="value"
        onCheckBoxClick={() => {}}
      />,
    );

    expect(wrapper.find('.tm-checkbox').exists()).toBe(true);
  });

  it('can apply the small class', () => {
    const wrapper = shallow(
      <CheckBox
        id="id"
        label="label"
        title="title"
        name="name"
        value="value"
        onCheckBoxClick={() => {}}
        small
      />,
    );

    expect(wrapper.find('.tm-checkbox-small').exists()).toBe(true);
  });

  it('can handle the onChange event', () => {
    const wrapper = shallow(
      <CheckBox
        id="id"
        label="label"
        title="title"
        name="name"
        value="value"
        onCheckBoxClick={() => {}}
      />,
    );

    const input = wrapper.find('input');
    input.simulate('change', { target: { checked: true } });
    expect(wrapper.instance().state.checked.value).toBe(false);
  });

  it('matches snapshot', () => {
    const wrapper = shallow(
      <CheckBox
        id="id"
        label="label"
        title="title"
        name="name"
        value="value"
        onCheckBoxClick={() => {}}
      />,
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('matches snapshot when small = true and labelSrOnly = true', () => {
    const wrapper = shallow(
      <CheckBox
        id="id"
        label="label"
        title="title"
        name="name"
        value="value"
        onCheckBoxClick={() => {}}
        small
        labelSrOnly
      />,
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
