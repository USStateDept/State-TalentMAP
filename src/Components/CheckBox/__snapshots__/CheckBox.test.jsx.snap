// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBoxComponent matches snapshot 1`] = `
<div
  className="usa-grid-full  tm-checkbox "
>
  <input
    checked="value"
    disabled={false}
    id="id"
    name="name"
    onChange={[Function]}
    title="title"
    type="checkbox"
    value="value"
  />
  <label
    htmlFor="id"
  >
    <span
      className=""
    >
      label
    </span>
  </label>
</div>
`;

exports[`CheckBoxComponent matches snapshot when small = true and labelSrOnly = true 1`] = `
<div
  className="usa-grid-full  tm-checkbox tm-checkbox-small"
>
  <input
    checked="value"
    disabled={false}
    id="id"
    name="name"
    onChange={[Function]}
    title="title"
    type="checkbox"
    value="value"
  />
  <label
    htmlFor="id"
  >
    <span
      className="usa-sr-only"
    >
      label
    </span>
  </label>
</div>
`;
